####################################################
##
## author: sisichen
## date: 07/22/15
##
## This script plots a histogram of the start sites 
## generated by findStarts.py
####################################################

library(ggplot2)
library(plyr)

###############################
## Set User Parameters Here:
###############################

#startspath<-"~/Documents/RNAseqdata/DS-2/starts/"
#startspath<-"~/Documents/RNAseqdata/DS-4/sample3/starts/"
#startspath<-"~/Documents/RNAseqdata/DS-2/starts/"
startspath<-"~/Documents/RNAseqdata/DS4-IH/50bpstarts/"
#startspath<-"~/Documents/RNAseqdata/DS-4/starts/"

#startsname<-'DS2_AF'
#startsname<-'DS4-S3-crispr'
#startsname<-'DS2_hygroES_noXS'
#startsname<-'IH50'
#startsname<-'DS4_all_filtered'
startsname<-'IH50_refset'
#startsname<-'DS4_refset'

#TD_path<-'~/Documents/reference_genomes/hygroES/'
#TD_path<-'~/Documents/reference_genomes/crispr/'
TD_path<-'~/Documents/reference_genomes/mm9/Transcriptome/'

#TD_file<-'hygroES_data.txt'
#TD_file<-'crispr_data.txt'
TD_file<- 'mouse_transcripts_data.txt'

TD_fullpath<-paste(TD_path,TD_file, sep='')

#Specify gene names (which also sets ordering)
#geneNames<-c('KLF4','POU5F1', 'VIM', 'SOX2', 'ZFP42', 'GAPDH')
#geneNames<- c("HMOX2","GTF3A","TIMM17B","BMS1","ATAD3A","SRSF4","KHSRP","NUP50","PTGR1","GM10069","ZFP780B","RENBP","GADD45GIP1","NOB1","SNX5","X2310008H09RIK","MFAP1A","MRPL43","BC003965","IFITM2")
#geneNames<-'all'
geneNames<-c('GAPDH','ACTB', 'PPIA', 'POU5F1', 'SOX2', 'MYC', 'ZFP42', 'NANOG')


###############################
## Start processing the data
###############################

# load data on alignment starts made using findstarts.py
setwd(startspath)
startsdf <- read.table(paste(startsname,'_','startsmatrix.txt', sep=''), sep='\t')
genesdf <- read.table(paste(startsname,'_','startsgenes.txt', sep=''))
ngenes <- dim(startsdf)[2]
colnames(startsdf)<-genesdf$V1

# load transcriptome data
TDdf<-read.table(TD_fullpath, header=TRUE,stringsAsFactors = FALSE)
TDdf[,2]<-toupper(TDdf[,2])

# reorder columns according to order in TDdf
TDinds<-sort(match(colnames(startsdf), TDdf$NM_name))
TDkeep<-TDdf[TDinds,] # keep transcriptome data corresponding to our genes
col_order<-match(TDdf$NM_name, colnames(startsdf))
startsdf<-startsdf[,col_order]

if (!(geneNames=='all')) {
  # only keep genes whose names are within the list. 
  TDkeep<-TDdf[match(geneNames,TDdf$geneID),]
  inds<-match(TDkeep$NM_nam, genesdf$V1)
  startsdf<-startsdf[,inds]
} else{
  TDkeep<-TDdf
}

# the following function takes in a column of starts, 
# with the associated gene name and outputs a data frame 
# with the starts in one column and the gene name in the
# second column

removezeros <- function(x,genename) {
  zeros_ind<-which(x==0)
  
  if (length(zeros_ind)){
    crop_ind <- min(zeros_ind)
  } else {
    crop_ind <- length(x)
  }
  
  # cut vector at the crop index
  start = x[1:crop_ind-1]
  print(length(start))
  NM_name = rep(genename,length(start))
  data.frame(start,NM_name)
}

reshapeddf <- data.frame(); # initialize a data frame

for (i in 1:ngenes){
  newdf = removezeros(startsdf[,i], colnames(startsdf)[i])
  # subtract cropped data from total lengths
  #newdf$start<-totallengths[i]-newdf$start
  reshapeddf = rbind(reshapeddf, newdf)
  }

facet_labeller<- function(variable, value){ 
  print(value)
  inds<-match(value,TDdf$NM_name)
  TDdf[inds,'geneID']
  }

# Plot the transcription start sites: 

theme_set(theme_bw(10))

colorset<-brewer.pal(n=3,name ='Set1')

maxBP<-max(TDkeep$length)
linew<-1.25

ssplot<-ggplot(reshapeddf,aes(start)) + 
  facet_grid(NM_name~., scales="free_y", labeller=facet_labeller) + 
  geom_histogram(binwidth=30) +
  geom_vline(aes(xintercept=cdsStart), data=TDkeep, colour=colorset[3], size=linew) + 
  geom_vline(aes(xintercept=cdsEnd), data=TDkeep, colour=colorset[1], size=linew) + 
  geom_vline(aes(xintercept=length), data=TDkeep, colour=colorset[2], size=linew) + 
  scale_x_continuous(limits=c(0,maxBP), breaks=seq(0,maxBP,500))
print(ssplot)
ssname<-paste(startsname, '_', 'hist.png', sep='')
ggsave(file=ssname, width=8,height=6)




